msal/src/Microsoft.Identity.Client/Internal/Requests/RequestValidationHelper.cs

```diff
@@ -0,0 +1,71 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+using Microsoft.Identity.Core;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Identity.Client.Internal.Requests
+{
+    internal static class RequestValidationHelper
+    {
+#if DESKTOP || NETSTANDARD1_3
+        /// <summary>
+        /// Determines whether or not the cached client assertion can be used again for the next authentication request by checking it's
+        /// values against incoming request parameters.
+        /// </summary>
+        /// <param name="clientAssertionParameters">Incoming client assertion parameters containing cached client assertion</param>
+        /// <returns>Returns true if the previously cached client assertion is valid</returns>
+        public static bool ValidateClientAssertion(AuthenticationRequestParameters clientAssertionParameters)
+        {
+            if (clientAssertionParameters.ClientCredential == null)
+            {
+                throw new ArgumentException("The " + nameof(clientAssertionParameters) + " does not contain the ClientCredential");
+            }
+            else if (String.IsNullOrWhiteSpace(clientAssertionParameters.ClientCredential.Assertion))
+            {
+                return false;
+            }
+
+            //Check if all current client assertion values match incoming parameters and expiration time
+            //The clientCredential object contains the previously used values in the cached client assertion string
+            var clientCredential = clientAssertionParameters.ClientCredential;
+            var expired = (clientCredential.ValidTo <=
+                                        Jwt.JsonWebToken.ConvertToTimeT(DateTime.UtcNow +
+                                                                        TimeSpan.FromMinutes(
+                                                                        Constants
+                                                                        .ExpirationMarginInMinutes)));
+
+            return !expired && clientCredential.Audience == clientAssertionParameters.Authority.SelfSignedJwtAudience
+                && clientCredential.ContainsX5C == clientAssertionParameters.SendCertificate;
```

Comment:
I recommend putting this on a separate line with a meaningful variable name which can help reveal the intent.

Reply:
#Resolved